src/compiler/emitter.ts

```diff
@@ -144,81 +147,193 @@ module ts {
+                result.text = name;
+                return result;
             }
 
-            function isExistingName(location: Node, name: string) {
-                // check if resolver is aware of this name (if name was seen during the typecheck)
-                if (!resolver.isUnknownIdentifier(location, name)) {
-                    return true;
-                }
 
-                // check if name is present in generated names that were introduced by the emitter
-                if (currentScopeNames && hasProperty(currentScopeNames, name)) {
-                    return true;
+            function recordGeneratedName(name: string, isTempVariableName: boolean) {
+                if (!generatedNames) {
+                    generatedNames = {};
                 }
 
-                // check generated names in outer scopes
-                // let x;
-                // function foo() {
-                //    let x; // 1
-                //    function bar() {
-                //        {
-                //            let x; // 2
-                //        }
-                //        console.log(x); // 3
-                //    }
-                //}
-                // here both x(1) and x(2) should be renamed and their names should be different
-                // so x in (3) will refer to x(1)
-                let frame = lastFrame;
-                while (frame) {
-                    if (hasProperty(frame.names, name)) {
-                        return true;
+                generatedNames[name] = name;
+
+                if (isTempVariableName) {
+                    if (!currentScopeTemp.names) {
+                        currentScopeTemp.names = {};
                     }
-                    frame = frame.previous;
+                    currentScopeTemp.names[name] = name;
                 }
-                return false;
+
+                return name;
+            }
+
+            function recordTempDeclaration(name: Identifier): void {
+                if (!currentScopeTemp.tempVariables) {
+                    currentScopeTemp.tempVariables = [];
+                }
+                currentScopeTemp.tempVariables.push(name);
+            }
+
+            function createAndRecordTempVariable(location: Node, preferredName?: string): Identifier {
+                let temp = createTempVariable(location, preferredName);
+                recordTempDeclaration(temp);
+                return temp;
+            }
+
+
+            function assignGeneratedName(node: Node, name: string): string {
+                if (!nodeToGeneratedName) {
+                    nodeToGeneratedName = [];
+                }
+                return nodeToGeneratedName[resolver.getNodeId(node)] = unescapeIdentifier(name);
+            }
+
+            function generateNameForFunctionOrClassDeclaration(node: Declaration): string {
+                if (!node.name) {
+                    return assignGeneratedName(node, makeUniqueName("default"));

```

Comment:
"default" is not a good name for a function or class in general. Only if it is export default.